server/src/rooms/game/events/index.ts

```diff
@@ -345,7 +357,18 @@ gameEventDeserializer.register(EnteredDiscardPhase);
 
 export class EnteredDefeatPhase extends KindOnlyGameEvent {
   apply(game: GameState): void {
-    game.phase = Phase.defeat;
+    if(game.upkeep <= 0){
```

Comment:
I'm not sure I'm comfortable with always adding the defeat phase and checking inside it to see if we should be in the defeat phase.. perhaps this is a naming issue and I'd be fine if it were called `CleanupPhase` or `VerifyPhase` or something along those lines. Other thoughts?

Reply:
So the reason for this is because of the way the events get applied. The finalize methods and the defeat methods are applied in the same sweep. So, for example, sandstorm drops upkeep in its finalize method, but the defeat command is evaluated before the drop is applied, making it possible for upkeep to be negative. By always applying defeat and moving the logic to the event, we can always get the most recent upkeep, so it works. I hope that’s clear, but if it’s not @cpritcha might be able to explain better.

Reply:
My concern has more to do with naming and clarity of the codebase. We're not entering the defeat phase, we are checking whether the group is still alive and kicking.

I'd like for us to have ELI5-able semantics on how we are applying system health contributions and decrements and when they are checked. @cpritcha @chrstngyn feel free to chime in though if we want to punt on this for tomorrow's pretest since the other changes look good to me and are a clear improvement over the nothing that's there now.